content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,526 @@
+
+This output shows that, inside of `slow_function()`, `step_1()` took about a third of the total time and `step_2()` took two-thirds. All of the time in both of those functions was due to calling `pause()`.
+
+Profiling should be your first line of defense against slow-running code. Often, profiling will surface slowdowns in unexpected places, and solutions to address those slowdowns may have little to do with usage of tidy tools. To learn more about profiling, the [Measuring performance](https://adv-r.hadley.nz/perf-measure.html) chapter in Hadley Wickham's book [Advanced R](https://adv-r.hadley.nz/index.html) is a great place to start.
+
+### bench
+
+profvis is a powerful tool to surface code slowdowns. Often, though, it may not be immediately clear how to _fix_ that slowdown. The bench package allows users to quickly test out how long different approaches to solving a problem take.
+
+For example, say we want to take the sum of the numbers in a list, but we've identified via profiling that this operation is slowing our code down:
+
+```{r numbers}
+numbers <- as.list(1:5)
+
+numbers
+```
+
+One approach could be using the `Reduce()` function:
+
+```{r reduce}
+Reduce(sum, numbers)
+```
+
+Another could involve converting to a vector with `unlist()` and then using `sum()`:
+
+```{r sum-unlist}
+sum(unlist(numbers))
+```
+
+You may have some other ideas of how to solve this problem! How do we figure out which one is fastest, though? The `bench::mark()` function from bench takes in different proposals to solve the same problem and returns a tibble with information about how long they took (among other things.)
+
+```{r bm-reduce}
+res <-
+  bench::mark(
+    approach_1 = Reduce(sum, numbers),
+    approach_2 = sum(unlist(numbers))
+  )
+
+res %>% select(expression, median)
+```
+
+The other nice part about `bench::mark()` is that it will check that each approach gives the same output, so that you don't mistakenly compare apples and oranges.
+
+```{r median-time, include = FALSE}
+median_time <- res$median[[1]]
+```
+
+There are two important lessons to take in from this output:
+
+* The `sum(unlist())` approach was wicked fast compared to `Reduce()`. 
+* Both of these expressions were fast. Even the slower of the two took `r format(median_time)`---to put that in perspective, that expression could complete `r format(round(1 / as.numeric(median_time)))` iterations in a second! Keeping this bigger picture in mind is always important when benchmarking; if code runs fast enough to not be an issue in practical situations, then it need not be optimized in favor of less readable or safe code.
+
+The results of little experiments like this one can be surprising at first. Over time, though, you will develop intuition for the fastest way to solve problems you commonly solve, and will write fast code the first time around! 
```

Comment:
It might be helpful to say something about the "intuition" in this case. For me, it goes something like:

> In this case, using `Reduce()` means calling `sum()` many times, approximately once for each element of the list, and while `sum()` isn't particularly slow, calling an R function many times tends to have non-negligible overhead. Whereas with the `sum(unlist())` approach, there are only 2 R function calls - one for `unlist()` and one for `sum()`, which both immediately drop into C code.